src/genotyper.cc

```diff
@@ -41,6 +41,286 @@ static bool is_pseudo_ref_record(const bcf_hdr_t* hdr, bcf1_t* record) {
+    }
+
+    /// Given a format field for an input sample (unmapped_i), and
+    /// the unmapped_j-th value for this sample, find the corresponding
+    /// index of the this format field in the output.
+    /// Returns a negative value if this value cannot be mapped to the output
+    /// (e.g. allele-specific info for a trimmed allele), and raises error
+    /// if the sample cannot be mapped
+    int get_out_ind_of_value(int unmapped_i, int unmapped_j,
+                             const map<int, int>& sample_mapping,
+                             const vector<int> allele_mapping) {
+        int mapped_i = sample_mapping.at(unmapped_i);
+
+        // Sample should always be mappable
+        assert (mapped_i >= 0);
+
+        switch (field_info.number){
+            case RetainedFieldNumber::ALT:
+            case RetainedFieldNumber::ALLELES:
+            {
+                // Fall through for both cases that require allele_mapping
+                int mapped_j = allele_mapping[unmapped_j];
+                if (mapped_j < 0) {
+                    // Allele is not mappable (trimmed or is a gvcf record)
+                    return -1;
+                } else {
+                    return mapped_i * count + mapped_j;
+                }
+                break;
+            }
+            case RetainedFieldNumber::GENOTYPE:
+            {
+                // TODO: Generate mapping for PL field (Number=G)
+                return -1;
+            }
+            default:
+            {
+                // RetainedFieldNumber::BASIC case
+                return mapped_i * count + unmapped_j;
+            }
+        }
+    }
+
+public:
+
+    FormatFieldHelper(const retained_format_field field_info_, int n_samples_, int count_) : IFormatFieldHelper(field_info_, n_samples_, count_) {
+
+        switch (field_info.combi_method) {
+            case FieldCombinationMethod::MIN:
+                combine_f = FormatFieldHelper::min_element_wrapper;
+                break;
+            case FieldCombinationMethod::MAX:
+                combine_f = FormatFieldHelper::max_element_wrapper;
+                break;
+        }
+
+        for (int i=0; i<n_samples_ * count_; i++){
+            // We keep 1 vector for each eventual output value
+            // So if every sample contains X values (in vcf specification, Number=X), then we keep X vectors per sample.
+            format_v.push_back(vector<T>());
+        }
+    }
+
+    virtual ~FormatFieldHelper() = default;
+
+
+    Status add_record_data(const string& dataset, const bcf_hdr_t* dataset_header, bcf1_t* record,
+                           const map<int, int>& sample_mapping, const vector<int> allele_mapping,
+                           const vector<string>& field_names={}, int n_val_per_sample=-1) {
+
+        bool found = false;
+        if (n_val_per_sample < 0) {
+            n_val_per_sample = expected_n_val_per_sample(record);
+        }
+
+        vector<string> names_to_search = field_names;
```

Comment:
could make name_to_search a pointer, to avoid copying the vector on each invocation